presto-main/src/main/java/com/facebook/presto/operator/scalar/KleinbergBurstDetectionFunction.java

```diff
@@ -0,0 +1,219 @@
+        double timeLength = DOUBLE.getDouble(timeSeries, numEntries - 1) - DOUBLE.getDouble(timeSeries, 0);
+        checkCondition(timeLength > 0, INVALID_FUNCTION_ARGUMENT, "time length of the whole series must be greater than 0");
+
+        // The upper limit of burst levels [K].
+        int burstLevelUpperLimit = (int) (Math.ceil(1 + Math.log(timeLength / delta) / Math.log(s)));
+        checkArgument(burstLevelUpperLimit > 0);
+
+        double[] alpha = new double[burstLevelUpperLimit];  // Event rate at each burst level.
+        double[] alphaLn = new double[burstLevelUpperLimit];
+
+        alpha[0] = numEntries / timeLength;  // Average event rate.
+        alphaLn[0] = Math.log(alpha[0]);
+
+        for (int i = 1; i < burstLevelUpperLimit; i++) {
+            alpha[i] = s * alpha[i - 1];
+            alphaLn[i] = Math.log(alpha[i]);
+        }
+
+        // Initialize.
+        Node[] q = new Node[burstLevelUpperLimit];
+        double[] costs = new double[burstLevelUpperLimit];
+        Arrays.fill(costs, Double.MAX_VALUE);
+        costs[0] = 0;
+
+        // Start optimization.
+        double gammaLogN = gamma * Math.log((double) numEntries);
+        for (double interval : intervals) {
+            Node[] qNew = new Node[burstLevelUpperLimit];
+            double[] costsNew = new double[burstLevelUpperLimit];
+            for (int i = 0; i < burstLevelUpperLimit; i++) {
+                double[] c = new double[burstLevelUpperLimit];
+                for (int j = 0; j < burstLevelUpperLimit; j++) {
+                    c[j] = costs[j] + tau(gammaLogN, j, i);
+                }
+                int jMin = findIndexOfSmallest(c);
+                // Store the cost for setting the burst level i.
+                costsNew[i] = -logf(alpha, alphaLn, i, interval) + c[jMin];
+                qNew[i] = new Node(i, q[jMin]);
+            }
+            q = qNew;
+            costs = costsNew;
+        }
+
+        // Build output structure.
+        int[] bursts = new int[numEntries];
+        int seqMin = findIndexOfSmallest(costs);
+        int count = 0;
+        Node p = q[seqMin];
+        while (p != null) {
+            count++;
+            bursts[numEntries - count] = p.getState();
+            p = p.getParent();
+        }
+
+        BlockBuilder burstsBuilder = INTEGER.createBlockBuilder(null, numEntries);
+        for (int i = 0; i < numEntries; i++) {
+            burstsBuilder.writeInt(bursts[i]);
+        }
+
+        return burstsBuilder.build();
+    }
+
+    // Internal helper structure to hold discovered burst nodes.
+    private static class Node
+    {
+        private int state; // burst level: 0, 1, ..., K-1
+        private Node parent;
+
+        int getState()
+        {
+            return state;
+        }
+
+        Node getParent()
+        {
+            return parent;
+        }
+
+        Node(int state, Node parent)
+        {
+            this.state = state;
+            this.parent = parent;
```

Comment:
Check for null

Reply:
We expect null 'parent' since this is a self referential class.